src/librustdoc/passes/check_custom_code_classes.rs

```diff
@@ -0,0 +1,89 @@
+//! NIGHTLY & UNSTABLE CHECK: custom_code_classes_in_docs
+//!
+//! This pass will produce errors when finding custom classes outside of
+//! nightly + relevant feature active.
+
+use super::{span_of_attrs, Pass};
+use crate::clean::{Crate, Item};
+use crate::core::DocContext;
+use crate::fold::DocFolder;
+use crate::html::markdown::{find_testable_code, ErrorCodes, LangString};
+
+use rustc_session::parse::feature_err;
+use rustc_span::symbol::sym;
+
+crate const CHECK_CUSTOM_CODE_CLASSES: Pass = Pass {
+    name: "check-custom-code-classes",
+    run: check_custom_code_classes,
+    description: "check for custom code classes while not in nightly",
+};
+
+crate fn check_custom_code_classes(krate: Crate, cx: &DocContext<'_>) -> Crate {
+    let mut coll = CustomCodeClassLinter::new(cx);
+
+    coll.fold_crate(krate)
+}
+
+struct CustomCodeClassLinter<'a, 'tcx> {
+    cx: &'a DocContext<'tcx>,
+}
+
+impl<'a, 'tcx> CustomCodeClassLinter<'a, 'tcx> {
+    fn new(cx: &'a DocContext<'tcx>) -> Self {
+        Self { cx }
+    }
+}
+
+impl<'a, 'tcx> DocFolder for CustomCodeClassLinter<'a, 'tcx> {
+    fn fold_item(&mut self, item: Item) -> Option<Item> {
+        let dox = item.attrs.collapsed_doc_value().unwrap_or_default();
+
+        look_for_custom_classes(&self.cx, &dox, &item);
+
+        Some(self.fold_item_recur(item))
+    }
+}
+
+struct TestsWithCustomClasses {
+    custom_classes_found: Vec<String>,
+}
+
+impl crate::doctest::Tester for TestsWithCustomClasses {
+    fn add_test(&mut self, _: String, config: LangString, _: usize) {
+        self.custom_classes_found.extend(config.added_classes.into_iter());
+    }
+}
+
+crate fn look_for_custom_classes<'tcx>(cx: &DocContext<'tcx>, dox: &str, item: &Item) {
+    let hir_id = match cx.as_local_hir_id(item.def_id) {
+        Some(hir_id) => hir_id,
+        None => {
+            // If non-local, no need to check anything.
+            return;
+        }
+    };
```

Comment:
Hmm, I guess this is a policy question: if you re-export an item using `feature`s, should you also need to enable the feature in your own crate? @Manishearth what do you think?

Reply:
Feels like it would be very annoying for reexports, especially since this is docs-only. We can, however, _not render it_ if the feature isn't enabled.

Reply:
> We can, however, not render it if the feature isn't enabled.

Oh, I don't know if I like that. That means the documentation changes just by re-exporting it.

Reply:
I'm not sure how to resolve that, I mostly copied what was done in other passes when ignoring external items.

Reply:
> @jyn514 Does it? I assume the custom classes are only used by additional stylesheets, and reexports wouldn't have those stylesheets. So to me this is already broken.

So, Guillaume said

> The point here is to allow users to add a class themselves and disable the rustdoc highlighting so that they can then add a JS library which will do it.

It's unclear to me how people will actually be adding their own highlighting. Is the idea to use `--extend-css` for this? That would work regardless of the crate it gets re-exported in if you forget to pass it; and would work regardless of the crate if the class gets added.

> I'm not sure how to resolve that, I mostly copied what was done in other passes when ignoring external items.

You only use the HIR ID for lints, so depending on the policy we want you could either not warn at all, or warn on the `pub use` instead of the original item.

Reply:
Either `--extend-css` or `--html-in-header` and then pointing to hightlight.js or equivalent.

Reply:
Ok, so I think we should unconditionally apply the custom class, and not check the feature gate (because that breaks crates: https://github.com/rust-lang/rust/issues/82284).